### PR TITLE
Add node resource overrides

### PR DESCRIPTION
**Pull request by katrogan**
Signed-off-by: Katrina Rogan <katroganGH@gmail.com>

# TL;DR
Add node resource overrides

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flyte/issues/475
https://github.com/flyteorg/flyte/issues/1170

## Follow-up issue
_NA_

### CODE REVIEW

**Review by EngHabu — APPROVED** [review]

**Review by EngHabu — APPROVED** [review]

File: flytekit/core/node.py [hunk 1]
```diff
@@ -30,6 +32,7 @@ def __init__(
         self._flyte_entity = flyte_entity
         self._aliases: _workflow_model.Alias = None
         self._outputs = None
+        self._resources: Resources = None
```
Comment (wild-endeavor) [inline flytekit/core/node.py:35]: `typing.Optional[Resources]`?

File: flytekit/models/task.py [hunk 2]
```diff
@@ -100,8 +100,8 @@ def to_flyte_idl(self):
         :rtype: flyteidl.core.tasks_pb2.Resources
         """
         return _core_task.Resources(
-            requests=[r.to_flyte_idl() for r in self.requests],
-            limits=[r.to_flyte_idl() for r in self.limits],
+            requests=[r.to_flyte_idl() for r in self.requests] if self.requests is not None else None,
```
Comment (wild-endeavor) [inline flytekit/models/task.py:103]: we should revert this - locally at least, unit tests passed for me. https://github.com/flyteorg/flytekit/pull/525

File: flytekit/common/translator.py [hunk 3]
```diff
@@ -272,7 +273,9 @@ def get_serializable_node(
             inputs=entity.bindings,
             upstream_node_ids=[n.id for n in upstream_sdk_nodes],
             output_aliases=[],
-            task_node=workflow_model.TaskNode(reference_id=task_spec.template.id),
+            task_node=workflow_model.TaskNode(
+                reference_id=task_spec.template.id, overrides=TaskNodeOverrides(resources=entity._resources)
```
Comment (wild-endeavor) [inline flytekit/common/translator.py:277]: These aren't the Resource models though right?  They're the internal dataclass.  Don't they need to be translated?

File: flytekit/models/task.py (hunk 2 above)
Comment (katrogan) [reply flytekit/models/task.py:103]: but now that we can initialize the model with optionally empty requests or limits, we can't, right?

they fail for me

```
FAILED tests/flytekit/unit/models/core/test_workflow.py::test_workflow_template - TypeError: 'NoneType' object is not iterable
FAILED tests/flytekit/unit/models/core/test_workflow.py::test_task_node - TypeError: 'NoneType' object is not iterable
FAILED tests/flytekit/unit/models/core/test_workflow.py::test_node_task_with_no_inputs - TypeError: 'NoneType' object is not iterable
FAILED tests/flytekit/unit/models/core/test_workflow.py::test_node_task_with_inputs - TypeError: 'NoneType' object is not iterable
FAILED tests/flytekit/unit/models/core/test_workflow.py::test_branch_node - TypeError: 'NoneType' object is not iterable
```

File: flytekit/common/translator.py (hunk 3 above)
Comment (katrogan) [reply flytekit/common/translator.py:277]: they are already translated https://github.com/flyteorg/flytekit/pull/523/files#diff-b58fe30a139359c0e865665c57312b6fafd564875cd34e10f8a07189b5a90913R94

File: flytekit/core/node.py (hunk 1 above)
Comment (katrogan) [reply flytekit/core/node.py]: done, here and elsewhere

File: flytekit/common/translator.py (hunk 3 above)
Comment (wild-endeavor) [reply flytekit/common/translator.py:277]: oh i see.  can we update the type hint on line 36 of node.py then?

File: flytekit/models/task.py (hunk 2 above)
Comment (wild-endeavor) [reply flytekit/models/task.py:107]: then why is this passing? https://github.com/flyteorg/flytekit/pull/525

File: flytekit/common/translator.py (hunk 3 above)
Comment (katrogan) [reply flytekit/common/translator.py:277]: ah yes, thanks for the catch

File: flytekit/models/task.py [hunk 4]
```diff
@@ -69,7 +69,11 @@ def from_flyte_idl(cls, pb2_object):
             """
             return cls(name=pb2_object.name, value=pb2_object.value)
 
-    def __init__(self, requests, limits):
+    def __init__(
+        self,
+        requests: typing.Optional[typing.List[ResourceEntry]] = None,
```
Comment (wild-endeavor) [inline flytekit/models/task.py:74]: nix this too?

**Review by wild-endeavor — APPROVED** [review]
i meant just the optional part but cool, doesn't matter, already in the docstring anyways

**Pull request merged**